### PR TITLE
Google takeout

### PR DESCRIPTION
**Pull request by adriangibanelbtactic**
This pull request adds Google Takeout support with three defaults languages: English, Spanish and Catala.

It requires the final user to install imapclient python module so that imap_utf7 encode function can be used.

Do you think there's something to polish?

Maybe you don't like some of my additional commits (non related to Google Takeout).

Thank you for your feedback!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -100,8 +101,64 @@ Some email providers use alternative IMAP folder separators (for example, Hetzne
...
+python imap_upload.py --ssl --user=login@example.net --password=MyS3cr3t --host=mail.example.net --port=993 --error='All mail Including Spam and Trash_errors.mbox' --google-takeout 'All mail Including Spam and Trash.mbox'
+```
+
 For more details, please refer to the --help message:
 
 ```sh
 python imap_upload.py --help
 ```
+
+### Usage
```
Comment (rgladwell) [inline README.md:115]: Thanks for adding this, very useful.

File: README.md [hunk 2]
```diff
@@ -21,7 +21,8 @@ IMAP Upload is a tool for uploading a local mbox file to IMAP4 server. The most
 
 ### Requirements
 
-*   Python 2.5 or later.
+*   Python 3.5 or later.
+*   imapclient ( `pip3 install imapclient` )
```
Comment (rgladwell) [inline README.md:25]: A bit cautious about adding another dependency, but I agree the default Python IMAP is insufficient.

File: imap_upload.py [hunk 3]
```diff
@@ -133,6 +152,14 @@ def parse_args(self, args):
...
             for (k, v) in dest.__dict__.items():
                 setattr(options, k, v)
+        if ((options.google_takeout_box_as_base_folder) and (not (options.google_takeout))):
+            self.error("--google-takeout-box-as-base-folder needs --google-takeout option")
+        if ((options.google_takeout_first_label) and (not (options.google_takeout))):
+            self.error("--google-takeout-first-label needs --google-takeout option")
+        if ((options.google_takeout_language) and (not (options.google_takeout))):
+            self.error("--google-takeout-language needs --google-takeout option")
+        if (not (options.google_takeout_language in self.google_takeout_supported_languages)):
+            self.error("--google-takeout-language: '%s' is not a supported language. Supported languages: '%s'." % (options.google_takeout_language, " ".join(self.google_takeout_supported_languages)))
```
Comment (rgladwell) [inline imap_upload.py:162]: 👍

File: imap_upload.py [hunk 4]
```diff
@@ -214,18 +241,128 @@ def decode_header_to_string(header):
...
+                    labels_without_categories.append(labels[i])
+
+            labels = labels_without_categories
+
+            labels_without_special_imap_dirs = []
+            for i in range(len(labels)):
+                if (not (re.match(gmail_imap_str,labels[i]))):
+                    labels_without_special_imap_dirs.append(labels[i])
+
+            labels = labels_without_special_imap_dirs
```
Comment (rgladwell) [inline imap_upload.py:312]: This assignment seems to be overwritten by the assignment on line 318? `labels_without_special_imap_dirs` doesn't seem to be used by the code. Should we delete?

File: imap_upload.py [hunk 5]
```diff
@@ -246,23 +383,43 @@ def endAll(self):
...
+                if r != "OK":
+                    raise Exception(r2[0]) # FIXME: Should use custom class
+
             p.endOk()
             continue
         except socket.error as e:
             p.endNg("Socket error: " + str(e))
         except Exception as e:
-            p.endNg(e)
+            if debug:
```
Comment (rgladwell) [inline imap_upload.py:419]: Nice.

File: README.md (hunk 1 above)
Comment (adriangibanelbtactic) [reply README.md:115]: Yeah. I felt that the complete list of options was missing.
What better way of showing them than using the program's usage output?

File: imap_upload.py (hunk 5 above)
Comment (adriangibanelbtactic) [reply imap_upload.py:419]: I only used debug in the place where I needed for my own debugging. Feel free to extend it to other parts of the code if you want it useful for your own debugging.

File: README.md (hunk 2 above)
Comment (rgladwell) [reply README.md:25]: No, this change is great. We should convert all the IMAP code to this better library.

**Pull request merged**